internal/common/database/redis.go

```diff
@@ -0,0 +1,55 @@
+//go:generate mockgen -typed -source=redis.go -destination=redis_mock.go -package=database
+package database
+
+import (
+	"context"
+	"fmt"
+	"github.com/redis/go-redis/v9"
+	"log"
+	"time"
+)
+
+type RedisClientInterface interface {
+	Get(ctx context.Context, key string) (string, error)
+	Set(ctx context.Context, key string, value interface{}, expiration int) error
+	Delete(ctx context.Context, key string) error
+	Close() error
+}
+
+type RedisClient struct {
+	client redis.UniversalClient
+}
+
+func (c *RedisClient) Get(ctx context.Context, key string) (string, error) {
+	return c.client.Get(ctx, key).Result()
+}
+
+func (c *RedisClient) Set(ctx context.Context, key string, value interface{}, expiration int) error {
+	return c.client.Set(ctx, key, value, time.Duration(expiration)).Err()
+}
+
+func (c *RedisClient) Delete(ctx context.Context, key string) error {
+	return c.client.Del(ctx, key).Err()
+}
+
+func (c *RedisClient) Close() error {
+	return c.client.Close()
+}
+
+func NewRedisClient(addr, password string) *RedisClient {
+	client := &RedisClient{
+		client: redis.NewClient(&redis.Options{
+			Addr:     addr,
+			Password: password,
+		}),
+	}
+
+	// check connection
+	pong, err := client.client.Ping(context.Background()).Result()
+	if err != nil {
+		log.Fatalf("cailed to connect to Redis: %v", err)
+	}
```

Comment:
Just wondering sir ferry,
If the redis instance is not instantiated, should the entire application fails?
because I think it might make more sense for the application to still run since the caching is not technically the source of truth

Reply:
Thank you for raising this point. It’s a great observation! Whether or not the application should terminate on a failed Redis connection depends on how critical Redis is to the application’s functionality.

If Redis is being used as a cache (as you mentioned), and it’s not the primary source of truth, it would make sense to allow the application to continue running even if the Redis connection fails. In such a scenario, the application could log the connection issue as a warning and fallback to fetching data directly from the database or another source.

Reply:
Agreed, one way I can think of to handle this issue if the case does exist where redis is the primary source of truth, is having a lazy initialization on the dependency.

It means only to initialize the redis instance, if and only if it's necessary (could happen in runtime)